rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,834 @@
+        // out which files are referenced by valid manifests
+        self.process_manifests().await?;
+        // Then we scan all of the files and delete those that are
+        // not part of our valid set.
+        self.delete_unreferenced_files().await
+    }
+
+    fn validate(&self) -> Result<()> {
+        if (utc_now() - self.before) < Duration::days(MINIMUM_CLEANUP_DAYS) {
+            return Err(Error::invalid_input(format!(
+                "Cannot cleanup data less than {} days old",
+                MINIMUM_CLEANUP_DAYS
+            )));
+        }
+        Ok(())
+    }
+
+    async fn process_manifests(&self) -> Result<()> {
+        self.dataset
+            .object_store
+            .read_dir_all(&self.dataset.versions_dir(), None)
+            .await?
+            .try_filter(|path| future::ready(path.extension() == Some("manifest")))
+            .try_for_each_concurrent(
+                8,
+                |path| async move { self.process_manifest_file(&path).await },
+            )
+            .await
+    }
+
+    async fn process_manifest_file(&self, path: &Path) -> Result<()> {
+        // TODO: We can't cleanup invalid manifests.  There is no way to distinguish
+        // between an invalid manifest and a temporary I/O error.  It's also not safe
+        // to ignore a manifest error because if it is a temporary I/O error and we
+        // ignore it then we might delete valid data files thinking they are not
+        // referenced.
+
+        let manifest = read_manifest(&self.dataset.object_store, path).await?;
+        let dataset_version = self.dataset.version().version;
+        // Don't delete the latest version, even if it is old
+        let is_latest = dataset_version == manifest.version;
+        if is_latest || manifest.timestamp() >= self.before {
+            self.process_valid_manifest(&manifest, path).await?;
+            return Ok(());
+        } else {
+            let mut old_manifests = self.old_manifests.lock().unwrap();
+            old_manifests.push(path.clone());
+        }
+        Ok(())
+    }
+
+    async fn process_valid_manifest(
+        &self,
+        manifest: &Manifest,
+        manifest_path: &Path,
+    ) -> Result<()> {
+        // Read the indexes first as it is async and we want to avoid holding the referenced_files
+        // lock across an await point.
+        let indexes =
+            read_manifest_indexes(&self.dataset.object_store, manifest_path, manifest).await?;
+
+        let mut referenced_files = self.referenced_files.lock().unwrap();
+        for fragment in manifest.fragments.iter() {
+            for file in fragment.files.iter() {
+                let full_data_path = self.dataset.data_dir().child(file.path.as_str());
+                let relative_data_path = remove_prefix(&full_data_path, &self.dataset.base);
+                referenced_files.data_paths.insert(relative_data_path);
+            }
```

Comment:
🤔 I was going to say:

> this list might be large enough where we want to amortize the HashSet's calls to `Reserve`. If we pass an iterator to `Extend` it can use `size_hint()` on the iterator to resize before adding elements

```rust
            let data_paths = fragment.files.iter().map(|file| {
                let full_data_path = self.dataset.data_dir().child(file.path.as_str());
                remove_prefix(&full_data_path, &self.dataset.base)
            });
            referenced_files.data_paths.extend(data_paths);
```

**But** it looks like `extend()` works best when at least 50% of keys are new, which in many cases won't be true for us:
https://docs.rs/hashbrown/0.14.0/src/hashbrown/map.rs.html#6443-6458

So this is good as-is :+1:

Reply:
Ah, good point though, I hadn't considered it.  Yes, I imagine that most of the time the manifests will have a pretty similar list of fragments.